File: backend/api/schemas/advanced_ai.py
"""Pydantic models for the Advanced AI (Genetic Algorithm) endpoint."""

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class AdvancedAIRequest(BaseModel):
    """Optional Genetic Algorithm configuration body for /api/advanced-ai/run."""

    model_config = ConfigDict(extra="forbid")

    baseline_cluster_count: int = Field(default=3, ge=2, le=20)
    population_size: int = Field(default=8, ge=2, le=50)
    generations: int = Field(default=5, ge=1, le=50)
    mutation_rate: float = Field(default=0.2, ge=0.0, le=1.0)
    random_state: int = Field(default=42, ge=0)


class AdvancedAIResultResponse(BaseModel):
    """Structured GA optimization result."""

    model_config = ConfigDict(extra="ignore")

    runnable: bool
    status_message: str
    settings: dict[str, float] = Field(default_factory=dict)
    baseline_score: float | None = None
    best_score: float | None = None
    improvement_over_baseline: float | None = None
    best_config: dict[str, Any] | None = None
    generation_history: list[dict[str, Any]] = Field(default_factory=list)
    optimized_assignments: dict[str, int] = Field(default_factory=dict)
    optimized_cluster_sizes: dict[str, int] = Field(default_factory=dict)
    optimized_top_terms_per_cluster: dict[str, list[str]] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/advancedAi.ts
export interface AdvancedAIRequest {
  baseline_cluster_count?: number;
  population_size?: number;
  generations?: number;
  mutation_rate?: number;
  random_state?: number;
}

export interface AdvancedAIBestConfig {
  technology_group_count?: number;
  keyword_evidence_feature_limit?: number | string;
  keyword_phrase_upper_bound?: number;
  [key: string]: unknown;
}

export interface AdvancedAIGenerationEntry {
  generation: number;
  best_score: number | null;
  average_score: number | null;
  [key: string]: unknown;
}

export interface AdvancedAIResult {
  runnable: boolean;
  status_message: string;
  settings: Record<string, number>;
  baseline_score: number | null;
  best_score: number | null;
  improvement_over_baseline: number | null;
  best_config: AdvancedAIBestConfig | null;
  generation_history: AdvancedAIGenerationEntry[];
  optimized_assignments: Record<string, number>;
  optimized_cluster_sizes: Record<string, number>;
  optimized_top_terms_per_cluster: Record<string, string[]>;
  warnings: string[];
}

export const ADVANCED_AI_DEFAULTS: Required<AdvancedAIRequest> = {
  baseline_cluster_count: 3,
  population_size: 8,
  generations: 5,
  mutation_rate: 0.2,
  random_state: 42,
};

export const ADVANCED_AI_LIMITS = {
  baseline_cluster_count: { min: 2, max: 20, step: 1 },
  population_size: { min: 2, max: 50, step: 1 },
  generations: { min: 1, max: 50, step: 1 },
  mutation_rate: { min: 0, max: 1, step: 0.05 },
  random_state: { min: 0, max: 999_999, step: 1 },
} as const;
